### PR TITLE
feat: Add paginators and filter for transaction lists and collectibles

### PR DESCRIPTION
**Pull request by nuanyang233**
close [#2649 ](https://github.com/DimensionDev/Maskbook/issues/2649)
close [#2650 ](https://github.com/DimensionDev/Maskbook/issues/2650)
close [#2667 ](https://github.com/DimensionDev/Maskbook/issues/2667)

## Screenshots

image

image

### CODE REVIEW

File: packages/maskbook/src/plugins/Wallet/hooks/useCollectibles.ts [hunk 1]
```diff
@@ -30,3 +32,19 @@ export function useCollectiblesFromNetwork(address: string, provider: Collectibl
...
+    const values = useRef<Collectible[]>([])
+    return useAsyncRetry(async () => {
+        if (page === 1) values.current = []
+
+        if (!address) return []
+        const result = await WalletRPC.getAssetsListNFT(address.toLowerCase(), provider, page)
+        const erc721Tokens = await WalletRPC.getERC721TokensPaged(page, 50)
+
+        values.current.push(
+            ...uniqBy([...result, ...erc721Tokens.map((x) => ERC721TokenRecordToCollectible(x))], 'token_id'),
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Wallet/hooks/useCollectibles.ts:46]: Both `address` and `token_id` should be used to define an unique ERC721 token. Different tokens from different contracts may got same `token_id`.

File: packages/maskbook/src/plugins/Wallet/hooks/useTransactions.ts [hunk 2]
```diff
@@ -2,11 +2,23 @@ import { useAsyncRetry } from 'react-use'
...
+export function useTransactions(address: string, page?: number) {
+    const values = useRef<Transaction[]>([])
     const provider = useValueRef(currentPortfolioDataProviderSettings)
     return useAsyncRetry(async () => {
         if (!address) return []
-        return WalletRPC.getTransactionList(address.toLowerCase(), provider)
-    }, [address, provider])
+        if (page === 1) values.current = []
+
+        if (provider === PortfolioProvider.ZERION) {
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Wallet/hooks/useTransactions.ts:15]:
```typescript
switch (provider) {
    case PortfolioProvider.ZERION:
        break;
    case PortfolioProvider.DEBANK:
        break;
    default:
        unreachable(provider)
}
```

File: packages/maskbook/src/extension/options-page/DashboardComponents/WalletContent.tsx [hunk 3]
```diff
@@ -92,30 +94,53 @@ export const WalletContent = forwardRef<HTMLDivElement, WalletContentProps>(func
...
+            data-testid="delete_button">
             {t('delete')}
         </MenuItem>,
     )
 
+    const [transactionTypeMenu, openTransactionTypeMenu] = useMenu(
+        <MenuItem key="all" onClick={() => setTransactionType(FilterTransactionType.ALL)}>
+            {t('all_transactions')} {transactionType === FilterTransactionType.ALL ? <Check fontSize="small" /> : null}
+        </MenuItem>,
+        <MenuItem key="Send" onClick={() => setTransactionType(FilterTransactionType.SEND)}>
```
Comment (guanbinrui) [inline packages/maskbook/src/extension/options-page/DashboardComponents/WalletContent.tsx:134]: `key="Send"` => `key="Sent"`

File: packages/maskbook/src/extension/options-page/DashboardComponents/WalletContent.tsx [hunk 4]
```diff
@@ -92,30 +94,53 @@ export const WalletContent = forwardRef<HTMLDivElement, WalletContentProps>(func
...
+            {t('sent_transactions')}
+            {transactionType === FilterTransactionType.SEND ? <Check fontSize="small" /> : null}
```
Comment (guanbinrui) [inline packages/maskbook/src/extension/options-page/DashboardComponents/WalletContent.tsx:136]: Please add small gap between text and icon.

**Review by guanbinrui — CHANGES_REQUESTED** [review]

File: packages/maskbook/src/extension/options-page/DashboardComponents/WalletContent.tsx (hunk 3 above)
Comment (septs) [reply packages/maskbook/src/extension/options-page/DashboardComponents/WalletContent.tsx:134]: enum need to be rename?

```suggestion
        <MenuItem key="Sent" onClick={() => setTransactionType(FilterTransactionType.SENT)}>
```

File: packages/maskbook/src/plugins/Wallet/hooks/useCollectibles.ts (hunk 1 above)
Comment (septs) [reply packages/maskbook/src/plugins/Wallet/hooks/useCollectibles.ts:46]:
```suggestion
            ...uniqBy([...result, ...erc721Tokens.map(ERC721TokenRecordToCollectible)], 'token_id'),
```

File: packages/maskbook/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx [hunk 5]
```diff
@@ -1,11 +1,13 @@
...
+import { Box, Button, CircularProgress, makeStyles, Skeleton, Typography } from '@material-ui/core'
 import { CollectibleCard } from './CollectibleCard'
 import { createERC1155Token, createERC721Token } from '../../../../web3/helpers'
 import type { WalletRecord } from '../../../../plugins/Wallet/database/types'
 import { useChainId } from '../../../../web3/hooks/useChainState'
 import { formatEthereumAddress } from '../../../../plugins/Wallet/formatter'
 import { createContext } from 'react'
 import type { Collectible } from '../../../../plugins/Wallet/types'
+import AutoResize from 'react-virtualized-auto-sizer'
+import { FixedSizeGrid } from 'react-window'
```
Comment (Jack-Works) [inline packages/maskbook/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx:10]: Why you need `react-window` if you're using paginator?

File: packages/maskbook/src/plugins/Wallet/types/portfolio.ts [hunk 6]
```diff
@@ -25,6 +31,7 @@ export interface Transaction {
     failed: boolean
     pairs: TransactionPair[]
     gasFee: TransactionGasFee | undefined
+    transactionType: string
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Wallet/types/portfolio.ts:34]: Shouldn't it be `FilterTransactionType`?

File: packages/maskbook/src/plugins/Wallet/apis/opensea.ts [hunk 7]
```diff
@@ -113,12 +113,12 @@ export interface AssetsListResponse {
     assets: Asset[]
 }
 
-export async function getAssetsList(from: string, size = 50, page = 0) {
+export async function getAssetsList(from: string, page = 0, size = 50) {
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Wallet/apis/opensea.ts:116]:
```suggestion
export async function getAssetsList(from: string, opts: { page?: number, size?: number }) {
```

File: packages/maskbook/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx (hunk 5 above)
Comment (nuanyang233) [reply packages/maskbook/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx:10]: Due to part of the api interface, we can not correctly get the total of the data source, Paginator component can't correctly display the previous and next some state, so here we changed to use FixedList to achieve the effect of drop-down paging

File: packages/maskbook/src/plugins/Wallet/types/portfolio.ts (hunk 6 above)
Comment (nuanyang233) [reply packages/maskbook/src/plugins/Wallet/types/portfolio.ts:34]: It needs to be compatible with types from different data sources

File: packages/maskbook/src/extension/options-page/DashboardComponents/TransactionList/index.tsx [hunk 8]
```diff
@@ -69,12 +94,40 @@ export function TransactionList() {
...
+                            onItemsRendered={({ visibleStopIndex, overscanStopIndex }) => {
+                                if (dataSource.length === 0 || transactionsError || transactionsLoading) return
+
+                                if (
+                                    visibleStopIndex === overscanStopIndex &&
+                                    visibleStopIndex === dataSource.length - 1
+                                )
+                                    setPage((prev) => prev + 1)
+                            }}
+                            itemSize={96}
```
Comment (Jack-Works) [inline packages/maskbook/src/extension/options-page/DashboardComponents/TransactionList/index.tsx:111]: this doesn't seems right

**Review by Jack-Works — APPROVED** [review]

**Pull request merged**